### ISSUE TITLE
Scope of implied-do

### ISSUE THREAD

**Issue by gregbuchholz** [lfortran/lfortran#4099]
I'm not sure this is an issue, but it is different from gfortran and ifx.  I'm not sure if an implied-do is supposed to have a different scope for the variable.  Specifically this program:

```Fortran
program scope

    implicit none
    integer :: i
    real :: a(10)

    i = 999
    a = [((2.5*i), i=1,10)]

    print *, i

end program scope
```

...prints "999" with gfortran and ifx, but "10" with LFortran.

**Comment by certik** [comment]
Thanks for the report. I actually don't know either, we need to look into this. In general, by default we do not guarantee the value of the loop variable after the loop, but have a compiler option to guarantee the last iteration that you can use for legacy code.

For implied do loop as a user I would assume the loop variable `i`'s value also becomes "undefined". However, if the standard says the implied do loop runs in a different scope, then we need to fix it.

This reminds me of a relevant thread here: https://fortran-lang.discourse.group/t/implied-do-array-constructor-type-specs-and-differences-between-gfortran-intel-and-lfortran/8047 (I just noticed, was it started by you?). It seems that you can apparently declare variables in implied do loops, so that almost implies they have their own scope?

**Comment by gregbuchholz** [comment]
The relevant part of the [2023 standard](https://j3-fortran.org/doc/year/23/23-007r1.pdf) might be section 19.4:

>The name of a *data-i-do-variable* in a DATA statement or an *ac-do-variable* in an array constructor has a scope of its *data-implied-do* or *ac-implied-do*.

...and yes, I started that other thread about type specs on implied-do.  Wouldn't a `block` statement also shadow variables correctly?

```Fortran
program scope2

    implicit none
    integer :: i
    real :: a(10)

    i = 999
    block
        integer :: i
        i = 10
        print *, "in block, i: ", i
    end block

    print *, "after block, i: ", i

end program scope2

```

**Comment by certik** [comment]
I see. You can also declare a variable in the implied do loop (not supported in LFortran yet: #https://github.com/lfortran/lfortran/issues/4033), which would indeed seem equivalent to the block construct you wrote. However, if it is not declared, as in `    a = [((2.5*i), i=1,10)]`, is it equivalent to:
```fortran
    block
        integer :: i
        i = 10
        print *, "in block, i: ", i
    end block
```
or to
```
    block
        i = 10
        print *, "in block, i: ", i
    end block
```
which will overwrite the global variable?

**Comment by harperjf** [comment]
F2023 section 19.4 is not easy to follow. Metcalf et al. "Modern Fortran Explained" (2023 edition p.138) says 
```
The scope of the variable is the constructor-implied-do. Other statements, or even other
parts of the array constructor, may refer to another variable having the same name. The
value of the other variable is unaffected by execution of the array constructor and is available
outside the constructor-implied-do.
```
Exactly the same wording is in Metcalf et al. "Fortran 95/2003 Explained" (2004 edition p.122). That makes me think that  gfortran and ifx are right on this matter and lfortran has a bug.

**Comment by certik** [comment]
So probably this:
```fortran
integer :: i
i = 10
a = [((2.5*i), i=1,10)]
```
is equivalent to this:
```fortran
integer :: i
i = 10
block
    integer :: i
    do i = 1, 10
        a(i) = 2.5*i
    end do
end block
```

**Comment by kargl** [comment]
Yes, an implied-do index has the scope of the loop.  Yes, intel and gfortran give the expected conforming behavior.  Note, making an analogy to a `block` construct should be done with caution especially if you ignore `implicit none`.  See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=84386
